src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,216 @@
+/**
+ * @file ppo.hpp
+ * @author Xiaohong Ji
+ *
+ * This file is the implementation of PPO class, which implements
+ * proximal policy optimization algorithms.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_PPO_IMPL_HPP
+#define MLPACK_METHODS_RL_PPO_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "ppo.hpp"
+
+namespace mlpack {
+namespace rl {
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::PPO(TrainingConfig config,
+       ActorNetworkType actorNetwork,
+       CriticNetworkType criticNetwork,
+       PolicyType policy,
+       ReplayType replayMethod,
+       UpdaterType updater,
+       EnvironmentType environment):
+  config(std::move(config)),
+  actorNetwork(std::move(actorNetwork)),
+  criticNetwork(std::move(criticNetwork)),
+  updater(std::move(updater)),
+  policy(std::move(policy)),
+  replayMethod(std::move(replayMethod)),
+  environment(std::move(environment))
+{
+  if (actorNetwork.Parameters().is_empty())
+    actorNetwork.ResetParameters();
+  if (criticNetwork.Parameters().is_empty())
+    criticNetwork.ResetParameters();
+
+  this->updater.Initialize(actorNetwork.Parameters().n_rows,
+                           actorNetwork.Parameters().n_cols);
+  this->oldActorNetwork = actorNetwork;
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::Update()
+{
+  // Sample from previous experience.
+  arma::mat sampledStates;
+  std::vector<ActionType> sampledActions;
+  arma::colvec sampledRewards;
+  arma::mat sampledNextStates;
+  arma::icolvec isTerminal;
+  arma::colvec discountedRewards;
+
+  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+      sampledNextStates, isTerminal);
+
+  arma::mat nextActionValues;
+  double values;
+  criticNetwork.Predict(sampledNextStates, nextActionValues);
+  for (size_t i = sampledRewards.n_cols; i > 0; --i)
+  {
+    values = sampledRewards[i-1] + values * config.Discount();
+    discountedRewards.insert_rows(discountedRewards.n_cols, values);
```

Comment:
It should be more efficient, if we set the size of ``discountedRewards`` (outside the loop) first and use ``discountedRewards.col(i)`` afterwards.

Reply:
> It should be more efficient, if we set the size of `discountedRewards` (outside the loop) first and use `discountedRewards.col(i)` afterwards.

Sorry for miss this review, it is fixed.